src/proper.erl

```diff
@@ -1774,12 +1741,15 @@ shrink(Shrunk, [ImmInstance | Rest], {_Type,Prop}, Reason,
 	   Shrinks, ShrinksLeft, init, Opts);
 shrink(Shrunk, [RawImmInstance | Rest] = TestTail, {Type,Prop} = StrTest, Reason,
        Shrinks, ShrinksLeft, State, Opts) ->
-    ImmInstance = case proper_types:find_prop(user_nf, Type) of
-		      {ok, _} ->
-			  proper_gen:clean_instance(RawImmInstance);
-		      error ->
-			  RawImmInstance
-		  end,
+    ImmInstance = case proper_types:find_prop(is_user_nf, Type) of
+                    {ok, true} ->
+                      case proper_types:safe_is_instance(RawImmInstance, Type) of
+                        false -> proper_gen:clean_instance(RawImmInstance);
+                        true -> RawImmInstance
+                      end;
+                    %% Either {ok, false} or error, suggests no user nf.
```

Comment:
I think it is better that this code is written in one of the following two ways:
```erlang
    Other when Other =:= {ok, false} orelse Other =:= erorr -> RawImmInstance
```
or
```erlang
   {ok, false} -> RawImmInstance;
   error -> RawImmInstance
```
because then it is robust w.r.t. what `proper_types:find_prop` returns.